src/libcollections/range.rs

```diff
@@ -23,39 +23,75 @@ pub trait RangeArgument<T> {
     /// Start index (inclusive)
     ///
     /// Return start value if present, else `None`.
-    fn start(&self) -> Option<&T> {
+    fn start(&self) -> Option<T> {
         None
     }
 
     /// End index (exclusive)
     ///
     /// Return end value if present, else `None`.
-    fn end(&self) -> Option<&T> {
+    fn end(&self) -> Option<T> {
         None
     }
 }
 
-// FIXME add inclusive ranges to RangeArgument
-
 impl<T> RangeArgument<T> for RangeFull {}
 
-impl<T> RangeArgument<T> for RangeFrom<T> {
-    fn start(&self) -> Option<&T> {
-        Some(&self.start)
+impl<T: Copy> RangeArgument<T> for RangeFrom<T> {
+    fn start(&self) -> Option<T> {
+        Some(self.start)
     }
 }
 
-impl<T> RangeArgument<T> for RangeTo<T> {
-    fn end(&self) -> Option<&T> {
-        Some(&self.end)
+impl<T: Copy> RangeArgument<T> for RangeTo<T> {
+    fn end(&self) -> Option<T> {
+        Some(self.end)
     }
 }
 
-impl<T> RangeArgument<T> for Range<T> {
-    fn start(&self) -> Option<&T> {
-        Some(&self.start)
+impl<T: Copy> RangeArgument<T> for Range<T> {
+    fn start(&self) -> Option<T> {
+        Some(self.start)
+    }
+    fn end(&self) -> Option<T> {
+        Some(self.end)
     }
-    fn end(&self) -> Option<&T> {
-        Some(&self.end)
+}
+
+macro_rules! inclusive {
+    ($Int: ty) => {
+        impl RangeArgument<$Int> for RangeToInclusive<$Int> {
+            fn end(&self) -> Option<$Int> {
+                Some(self.end.checked_add(1).expect("inclusive range to maximum usize"))
+            }
+        }
+
+        impl RangeArgument<$Int> for RangeInclusive<$Int> {
+            fn start(&self) -> Option<$Int> {
+                match *self {
+                    RangeInclusive::Empty { at } => Some(at),
+                    RangeInclusive::NonEmpty { start, .. } => Some(start),
+                }
+            }
+            fn end(&self) -> Option<$Int> {
+                match *self {
+                    RangeInclusive::Empty { at } => Some(at),
+                    RangeInclusive::NonEmpty { end, .. } => {
+                        Some(end.checked_add(1).expect("inclusive range to maximum usize"))
```

Comment:
Hmm well it's easy enough to fix the message to `concat!("inclusive range to maximum ", stringify!($Int))` at least.

I see your point but I'm not sure I like privileging `usize`. But there's no reason to listen to me! I also understand the desire for stabilization, of course, but it may be premature since I think that range syntax is going to become more flexible (@aturon keeps reminding me to write that RFC...).